### PR TITLE
Replace `internal::GRAIN_SIZE` by `grain_size` (parameter).

### LINKED ISSUE #53013: `TensorIterator::for_each` should use the provided `grain_size` parameter
## 🐛 Bug

Follow-up from #52043

The parallelization of some operators rely on the `grain_size` passed to  `TensorIterator::for_each`. However, this function uses `internal::GRAIN_SIZE` instead, hindering parallelization of said operations. Some of the operations affected: `mode`, `min`, `max`. This would enable distribution of to-be-processed slices among the available processors.

## Additional context

Line where `for_each` is called with `grain_size = 1` (used by `mode`, `min`, and `max`). They are never parallelized, unless there is more than `internal::GRAIN_SIZE` slices (output elements).

https://github.com/pytorch/pytorch/blob/e43ea227fe014c587e2141070bef4320cb07d147/aten/src/ATen/native/cpu/TensorCompareKernel.cpp#L64

Probably, line 685 should have been:

```c++
} else if (numel < grain_size || at::get_num_threads() == 1) { 
    return serial_for_each(loop, {0, numel}); 
} else { 
```

https://github.com/pytorch/pytorch/blob/e43ea227fe014c587e2141070bef4320cb07d147/aten/src/ATen/TensorIterator.cpp#L681-L692

### PR DESCRIPTION
**Pull request by ysiraichi**
Fixes #53013

### CODE REVIEW

**Comment by ysiraichi** [comment]
Here are some relevant benchmarks for `max` and `min` kernels. Both _before_ and _after_ versions were executed with `torch.set_num_threads(24)`.

<details>
<summary>Benchmark Program</summary>

```python
import timeit
import sys

fn = 'max' # change here to some other kernel
device = 'cpu'
dims = 2
times = 1000

dtypes = [
    'uint8',
    'int16',
    'int32',
    'int64',
    'float32',
    'float64',
]

shapes = [
    (1000, 1000),
    (2000, 2000),
    (4000, 4000),
    (8000, 8000),
]

# Warm up first!
timeit.timeit(
    setup = f"import torch; a = torch.randn((200, 200), device = 'cpu').to(torch.int)",
    stmt  = f"torch.{fn}(a, dim = 1)",
    number = times
)

for shape in shapes:
    for dtype in dtypes:
        for dim in range(dims):
            print(f"torch.{fn}(a, dim = {dim}) with shape {shape} and dtype torch.{dtype} {times} times")
            print(timeit.timeit(
                setup = f"import torch; torch.set_num_threads(24); a = torch.randn({shape}, device = '{device}').to(torch.{dtype})",
                stmt  = f"torch.{fn}(a, dim = {dim})",
                number = times
            ))
```
</details>

<details>
<summary>Max Kernel</summary>

| calls | kernel | shape        | dim | dtype   | before (s) | after (s) |
|:-----:|:------:|:------------:|:---:|:-------:|-----------:|----------:|
|  1000 | max    | (1000, 1000) |   0 | uint8   |       0.76 |      0.10 |
|  1000 | max    | (1000, 1000) |   1 | uint8   |       0.74 |      0.10 |
|  1000 | max    | (1000, 1000) |   0 | int16   |       0.78 |      0.14 |
|  1000 | max    | (1000, 1000) |   1 | int16   |       0.76 |      0.10 |
|  1000 | max    | (1000, 1000) |   0 | int32   |       0.79 |      0.25 |
|  1000 | max    | (1000, 1000) |   1 | int32   |       0.76 |      0.10 |
|  1000 | max    | (1000, 1000) |   0 | int64   |       0.76 |      0.31 |
|  1000 | max    | (1000, 1000) |   1 | int64   |       0.65 |      0.07 |
|  1000 | max    | (1000, 1000) |   0 | float32 |       0.71 |      0.33 |
|  1000 | max    | (1000, 1000) |   1 | float32 |       0.56 |      0.11 |
|  1000 | max    | (1000, 1000) |   0 | float64 |       0.85 |      0.12 |
|  1000 | max    | (1000, 1000) |   1 | float64 |       0.64 |      0.05 |
|  1000 | max    | (2000, 2000) |   0 | uint8   |       3.15 |      0.18 |
|  1000 | max    | (2000, 2000) |   1 | uint8   |       2.95 |      0.17 |
|  1000 | max    | (2000, 2000) |   0 | int16   |       6.64 |      0.59 |
|  1000 | max    | (2000, 2000) |   1 | int16   |       3.00 |      0.17 |
|  1000 | max    | (2000, 2000) |   0 | int32   |      12.12 |      0.62 |
|  1000 | max    | (2000, 2000) |   1 | int32   |       3.24 |      0.22 |
|  1000 | max    | (2000, 2000) |   0 | int64   |      15.16 |      1.03 |
|  1000 | max    | (2000, 2000) |   1 | int64   |       2.58 |      0.14 |
|  1000 | max    | (2000, 2000) |   0 | float32 |      13.01 |      0.67 |
|  1000 | max    | (2000, 2000) |   1 | float32 |       2.45 |      0.14 |
|  1000 | max    | (2000, 2000) |   0 | float64 |      14.69 |      0.70 |
|  1000 | max    | (2000, 2000) |   1 | float64 |       2.56 |      0.13 |
|  1000 | max    | (4000, 4000) |   0 | uint8   |      45.73 |      3.86 |
|  1000 | max    | (4000, 4000) |   1 | uint8   |      11.79 |      1.14 |
|  1000 | max    | (4000, 4000) |   0 | int16   |      52.57 |      4.19 |
|  1000 | max    | (4000, 4000) |   1 | int16   |      12.08 |      0.96 |
|  1000 | max    | (4000, 4000) |   0 | int32   |      64.12 |      5.70 |
|  1000 | max    | (4000, 4000) |   1 | int32   |      13.02 |      1.16 |
|  1000 | max    | (4000, 4000) |   0 | int64   |      76.38 |      6.93 |
|  1000 | max    | (4000, 4000) |   1 | int64   |      11.95 |      2.48 |
|  1000 | max    | (4000, 4000) |   0 | float32 |      59.80 |      5.52 |
|  1000 | max    | (4000, 4000) |   1 | float32 |       9.22 |      1.84 |
|  1000 | max    | (4000, 4000) |   0 | float64 |      71.23 |      6.89 |
|  1000 | max    | (4000, 4000) |   1 | float64 |      12.19 |      2.46 |
|  1000 | max    | (8000, 8000) |   0 | uint8   |     207.01 |     19.40 |
|  1000 | max    | (8000, 8000) |   1 | uint8   |      47.00 |      3.90 |
|  1000 | max    | (8000, 8000) |   0 | int16   |     231.69 |     22.30 |
|  1000 | max    | (8000, 8000) |   1 | int16   |      47.96 |      4.92 |
|  1000 | max    | (8000, 8000) |   0 | int32   |     277.71 |     28.90 |
|  1000 | max    | (8000, 8000) |   1 | int32   |      50.61 |      5.58 |
|  1000 | max    | (8000, 8000) |   0 | int64   |     388.01 |     44.01 |
|  1000 | max    | (8000, 8000) |   1 | int64   |      48.58 |     12.39 |
|  1000 | max    | (8000, 8000) |   0 | float32 |     374.22 |     32.54 |
|  1000 | max    | (8000, 8000) |   1 | float32 |      34.25 |     13.04 |
|  1000 | max    | (8000, 8000) |   0 | float64 |     387.24 |     42.09 |
|  1000 | max    | (8000, 8000) |   1 | float64 |      48.52 |     10.48 |
</details>

<details>
<summary>Min Kernel</summary>

|   calls | kernel   | shape          |   dim | dtype     |   before (s) |   after (s) |
| :-----: | :------: | :------------: | :---: | :-------: | -----------: | ----------: |
|    1000 | max      | (1000, 1000)   |     0 | uint8     |         0.75 |        0.10 |
|    1000 | max      | (1000, 1000)   |     1 | uint8     |         0.73 |        0.09 |
|    1000 | max      | (1000, 1000)   |     0 | int16     |         0.78 |        0.10 |
|    1000 | max      | (1000, 1000)   |     1 | int16     |         0.75 |        0.09 |
|    1000 | max      | (1000, 1000)   |     0 | int32     |         0.79 |        0.26 |
|    1000 | max      | (1000, 1000)   |     1 | int32     |         0.75 |        0.10 |
|    1000 | max      | (1000, 1000)   |     0 | int64     |         0.78 |        0.31 |
|    1000 | max      | (1000, 1000)   |     1 | int64     |         0.71 |        0.07 |
|    1000 | max      | (1000, 1000)   |     0 | float32   |         0.73 |        0.27 |
|    1000 | max      | (1000, 1000)   |     1 | float32   |         0.56 |        0.07 |
|    1000 | max      | (1000, 1000)   |     0 | float64   |         0.82 |        0.09 |
|    1000 | max      | (1000, 1000)   |     1 | float64   |         0.64 |        0.06 |
|    1000 | max      | (2000, 2000)   |     0 | uint8     |         2.99 |        0.17 |
|    1000 | max      | (2000, 2000)   |     1 | uint8     |         2.94 |        0.22 |
|    1000 | max      | (2000, 2000)   |     0 | int16     |        11.57 |        0.88 |
|    1000 | max      | (2000, 2000)   |     1 | int16     |         2.97 |        0.20 |
|    1000 | max      | (2000, 2000)   |     0 | int32     |        12.87 |        0.63 |
|    1000 | max      | (2000, 2000)   |     1 | int32     |         3.47 |        0.18 |
|    1000 | max      | (2000, 2000)   |     0 | int64     |        14.93 |        0.70 |
|    1000 | max      | (2000, 2000)   |     1 | int64     |         3.28 |        0.12 |
|    1000 | max      | (2000, 2000)   |     0 | float32   |        12.70 |        0.69 |
|    1000 | max      | (2000, 2000)   |     1 | float32   |         2.33 |        0.13 |
|    1000 | max      | (2000, 2000)   |     0 | float64   |        16.00 |        1.00 |
|    1000 | max      | (2000, 2000)   |     1 | float64   |         3.34 |        0.13 |
|    1000 | max      | (4000, 4000)   |     0 | uint8     |        45.52 |        4.20 |
|    1000 | max      | (4000, 4000)   |     1 | uint8     |        11.67 |        1.18 |
|    1000 | max      | (4000, 4000)   |     0 | int16     |        50.86 |        4.63 |
|    1000 | max      | (4000, 4000)   |     1 | int16     |        12.03 |        0.62 |
|    1000 | max      | (4000, 4000)   |     0 | int32     |        60.12 |        5.45 |
|    1000 | max      | (4000, 4000)   |     1 | int32     |        12.73 |        1.19 |
|    1000 | max      | (4000, 4000)   |     0 | int64     |        69.86 |        7.12 |
|    1000 | max      | (4000, 4000)   |     1 | int64     |        13.22 |        2.53 |
|    1000 | max      | (4000, 4000)   |     0 | float32   |        58.43 |        5.88 |
|    1000 | max      | (4000, 4000)   |     1 | float32   |         8.98 |        1.44 |
|    1000 | max      | (4000, 4000)   |     0 | float64   |        72.19 |        7.29 |
|    1000 | max      | (4000, 4000)   |     1 | float64   |        13.00 |        2.60 |
|    1000 | max      | (8000, 8000)   |     0 | uint8     |       199.76 |       18.87 |
|    1000 | max      | (8000, 8000)   |     1 | uint8     |        47.09 |        4.92 |
|    1000 | max      | (8000, 8000)   |     0 | int16     |       227.88 |       22.04 |
|    1000 | max      | (8000, 8000)   |     1 | int16     |        47.97 |        4.97 |
|    1000 | max      | (8000, 8000)   |     0 | int32     |       275.86 |       28.50 |
|    1000 | max      | (8000, 8000)   |     1 | int32     |        51.17 |        5.46 |
|    1000 | max      | (8000, 8000)   |     0 | int64     |       385.94 |       42.57 |
|    1000 | max      | (8000, 8000)   |     1 | int64     |        52.24 |       10.29 |
|    1000 | max      | (8000, 8000)   |     0 | float32   |       273.11 |       34.50 |
|    1000 | max      | (8000, 8000)   |     1 | float32   |        35.84 |       11.83 |
|    1000 | max      | (8000, 8000)   |     0 | float64   |       390.70 |       43.87 |
|    1000 | max      | (8000, 8000)   |     1 | float64   |        50.85 |       11.29 |
</details>

**Review by ngimel — APPROVED** [review]

**Comment by jeffdaily** [comment]
@ysiraichi and @ngimel, we have a few questions about this PR.  We are observing ~25% performance drops in detectron2 retinanet and maskrcnn benchmarks due to this PR.  Would this PR affect both CPU and GPU backends?  I understand that this PR is changing some pretty old code, and superficially it looks like it is fixing an old copy/paste mistake of not using the `grain_size` parameter passed in by the caller.  But was there any perf impact study done?  Can someone else observe the same perf regression?

**Comment by ngimel** [comment]
@jeffdaily this PR affects only operations performed on CPU, and only those operations that specify grain argument to `iter.for_each` (a few indexing operations and mode etc, as mentioned in the issue). We didn't observe perf regression on torchbench benchmarks (cc @xuzhao9).

**Comment by pruthvistony** [comment]
@ngimel,
We see performance improvement on ROCm, when we revert this PR for detectron2 workload.
I ran the same on CUDA and see performance improvement on CUDA too.

CUDA version – 11.3
CUDnn – 8
PyTorch - 1.10 version

[01/18 14:28:01 d2.utils.events]:  eta: 0:00:00  iter: 89999  total_loss: 0.5074  loss_cls: 0.2414  loss_box_reg: 0.2759  time: 0.2916  data_time: 0.0266  lr: 2.5e-05  max_mem: 4148M
[01/18 14:28:01 d2.engine.hooks]: Overall training speed: 89998 iterations in 7:17:26 (0.2916 s / it)
[01/18 14:28:01 d2.engine.hooks]: Total training time: 7:17:57 (0:00:30 on hooks)

PyTorch – 1.10 + PR #53177 revert
[01/17 12:19:01 d2.utils.events]:  eta: 0:00:00  iter: 89999  total_loss: 0.5159  loss_cls: 0.2456  loss_box_reg: 0.2618  time: 0.2264  data_time: 0.0097  lr: 2.5e-05  max_mem: 4148M
[01/17 12:19:01 d2.engine.hooks]: Overall training speed: 89998 iterations in 5:24:36 (0.2164 s / it)
[01/17 12:19:01 d2.engine.hooks]: Total training time: 5:24:36 (0:00:30 on hooks)

**Comment by ngimel** [comment]
This PR concerns only cpu. cc @ysiraichi

**Comment by ysiraichi** [comment]
Yes. As far as I know, all the kernels that call `for_each` live on CPU.
